### PR TITLE
Assorted Thorns fixes

### PR DESCRIPTION
**Pull request by legokidlogan**
- Fixes the long-standing issue with the TeslaHitboxes effect attaching to other entities
- Fixes thorns modifying damage info used by PostEntityTakeDamage hook listeners that come after it (e.g. hitmarkers)
- Thorns uses its own damage inflictor now, so it has a proper killicon and can prevent infinite reflection loops
- Fixes the zaps using an episodic sound
- Removes the spammy debug prints

### CODE REVIEW

File: lua/powerups/server/thorns.moon [hunk 1]
```diff
@@ -111,43 +105,56 @@ class ThornsPowerup extends BasePowerup
             return unless ent == @owner
             return if took == false
 
-            originalAttacker = dmg\GetAttacker!
-            return unless IsValid originalAttacker
-            --return unless originalAttacker\IsPlayer!
-            return if ent == originalAttacker
+            attacker = dmg\GetAttacker!
+            return unless IsValid attacker
+            --return unless attacker\IsPlayer!
```
Comment (sarahsturgeon) [inline lua/powerups/server/thorns.moon]:
```suggestion
```

File: lua/powerups/server/thorns.moon [hunk 2]
```diff
@@ -111,43 +105,56 @@ class ThornsPowerup extends BasePowerup
...
+            return if ent == attacker
+
+            inflictor = dmg\GetInflictor!
+            return if IsValid(inflictor) and inflictor\GetClass! == "cfc_powerup_thorns_inflictor"
 
             damageAmount = dmg\GetDamage!
 
             return unless damageAmount > 0
-            --return unless originalAttacker\Alive! -- TODO: Does this actually prevent reflect damage being reflected?
+            --return unless attacker\Alive! -- TODO: Does this actually prevent reflect damage being reflected?
```
Comment (sarahsturgeon) [inline lua/powerups/server/thorns.moon]:
```suggestion
```

File: lua/powerups/server/thorns.moon [hunk 3]
```diff
@@ -111,43 +105,56 @@ class ThornsPowerup extends BasePowerup
...
+                    \SetDamage reflectedAmount
+                    \SetDamageType DMG_GENERIC
 
-            originalAttacker\TakeDamageInfo dmg
+                    attacker\TakeDamageInfo refDmg
 
-            --originalAttacker\ChatPrint "[CFC Powerups] You took #{Round newDamageAmount} reflected damage!"
+            @QueueDamageForBroadcast attacker, reflectedAmount
+
+            --attacker\ChatPrint "[CFC Powerups] You took #{Round newDamageAmount} reflected damage!"
```
Comment (sarahsturgeon) [inline lua/powerups/server/thorns.moon]:
```suggestion
```

File: lua/powerups/server/thorns.moon [hunk 4]
```diff
@@ -111,43 +105,56 @@ class ThornsPowerup extends BasePowerup
...
         damageWatcher = @DamageWatcher!
         hook.Add "PostEntityTakeDamage", @HookName, damageWatcher
         hook.Add "DoPlayerDeath", @HookName, (ply, _, dmg ) ->
             damageWatcher ply, dmg
 
-        timer.Create @TimerName, duration, 1, -> @Remove!
-        timer.Create @ZapperName, 0.1, duration * 10, ->
-            Effect("TeslaHitboxes", @aoeEffect, true, true )
+        timer.Create @TimerName, @duration, 1, -> @Remove!
+        timer.Create @ZapperName, 0.1, @duration * 10, -> @PlayAoeEffect!
```
Comment (sarahsturgeon) [inline lua/powerups/server/thorns.moon:152]:
```suggestion
        timer.Create @TimerName, @duration, 1, @\Remove
        timer.Create @ZapperName, 0.1, @duration * 10, @\PlayAoeEffect
```

File: lua/powerups/server/thorns.moon [hunk 5]
```diff
@@ -159,6 +166,7 @@ class ThornsPowerup extends BasePowerup
     Refresh: =>
         super self
         timer.Start @TimerName
+        timer.Create @ZapperName, 0.1, @duration * 10, -> @PlayAoeEffect!
```
Comment (sarahsturgeon) [inline lua/powerups/server/thorns.moon:164]:
```suggestion
        timer.Create @ZapperName, 0.1, @duration * 10, -> @\PlayAoeEffect
```

**Review by sarahsturgeon — APPROVED** [review]

File: lua/powerups/server/thorns.moon (hunk 4 above)
Comment (legokidlogan) [reply lua/powerups/server/thorns.moon:152]: No.
image

**Pull request merged**